src/connector/src/source/mqtt/enumerator/mod.rs

```diff
@@ -0,0 +1,144 @@
+    type Properties = MqttProperties;
+    type Split = MqttSplit;
+
+    async fn new(
+        properties: Self::Properties,
+        context: SourceEnumeratorContextRef,
+    ) -> ConnectorResult<MqttSplitEnumerator> {
+        let (client, mut eventloop) = properties.common.build_client(context.info.source_id)?;
+
+        let topic = properties.common.topic.clone();
+        let mut topics = HashSet::new();
+        if !topic.contains('#') && !topic.contains('+') {
+            topics.insert(topic.clone());
+        }
+
+        client
+            .subscribe(topic.clone(), rumqttc::v5::mqttbytes::QoS::AtMostOnce)
+            .await?;
+
+        let cloned_client = client.clone();
+
+        let topics = Arc::new(RwLock::new(topics));
+
+        let connected = Arc::new(AtomicBool::new(false));
+        let connected_clone = connected.clone();
+
+        let stopped = Arc::new(AtomicBool::new(false));
+        let stopped_clone = stopped.clone();
+
+        let topics_clone = topics.clone();
+        tokio::spawn(async move {
+            while !stopped_clone.load(std::sync::atomic::Ordering::Relaxed) {
+                match eventloop.poll().await {
+                    Ok(Event::Outgoing(Outgoing::Subscribe(_))) => {
+                        connected_clone.store(true, std::sync::atomic::Ordering::Relaxed);
+                    }
+                    Ok(Event::Incoming(Incoming::Publish(p))) => {
+                        let topic = String::from_utf8_lossy(&p.topic).to_string();
+                        let exist = {
+                            let topics = topics_clone.read().await;
+                            topics.contains(&topic)
+                        };
+
+                        if !exist {
+                            let mut topics = topics_clone.write().await;
+                            topics.insert(topic);
+                        }
+                    }
+                    Ok(_) => {}
+                    Err(err) => {
+                        if let ConnectionError::Timeout(_) = err {
+                            continue;
+                        }
+                        tracing::error!(
+                            "[Enumerator] Failed to subscribe to topic {}: {}",
+                            topic,
+                            err.as_report(),
+                        );
+                        connected_clone.store(false, std::sync::atomic::Ordering::Relaxed);
+                        cloned_client
+                            .subscribe(topic.clone(), rumqttc::v5::mqttbytes::QoS::AtMostOnce)
+                            .await
+                            .unwrap();
+                    }
+                }
+            }
+        });
+
+        Ok(Self {
+            client,
+            topics,
+            topic: properties.common.topic,
+            connected,
+            stopped,
+        })
+    }
+
+    async fn list_splits(&mut self) -> ConnectorResult<Vec<MqttSplit>> {
```

Comment:
Split number (vec length) here stands for the max parallelism for each source streaming job.
Take kafka source as an example, each split matches a partition in concept so we can read multiple partitions in parallel. I am not sure what struct at the place in mqtt.

Reply:
Mqtt doesn't have the concept of a partition, but it does support hierarchies `/parent/1/alerts` `/parent/2/alerts` and you can use `/parent/+/alerts` which will have 2 splits (one per topic)